### PR TITLE
Removing clang warning in EgammaHLTPixelMatchVarProducer

### PR DESCRIPTION
**Pull request by Sam-Harper**
#### PR description:

This PR fixes the clang warning reported in https://github.com/cms-sw/cmssw/pull/46319

The warning was harmless but in looking at the code, figured would make it a bit more robust against bad data (and if we ever added a new pixel layer unlikely that it is)

So the code now explicitly verfies that the value is in range and throws an exception if caused. I also strongly suspect but did not prove 100% that numeric_limits::max is an invalid value for this specific quanity and put that as an exception as well.  I'm running over a bunch of data to verify this (about 20k HLTPhysics events and will try to run on an EGamma file)

Finally it should be noted that this code is only run if productsToWrite  = 2 which is only done for debugging when we wish to dump the values of this producer for studies. So it doesnt normally run in the HLT and I dont believe there is any workflow that runs this code. 
 
So this should be a zero output change. As this code does not run normally and when it does run the data should always pass these checks. 
 
#### PR validation:

Runs as expected on my local machine based on HLT timing data. Will run on E/gamma data later with productsToWrite=2 later.

### CODE REVIEW

File: RecoEgamma/EgammaHLTProducers/plugins/EgammaHLTPixelMatchVarProducer.cc [hunk 1]
```diff
@@ -41,7 +41,24 @@ namespace {
...
+
+      int layerOrDiskNr = seed.layerOrDiskNr(hitNr);
+      if (layerOrDiskNr == 0 || layerOrDiskNr > 4) {
+        if (layerOrDiskNr == std::numeric_limits<int>::max()) {
+          throw cms::Exception("LogicError")
+              << "The layerOrDiskNr of hitnr " << hitNr << " is " << layerOrDiskNr
+              << " which is equal to numeric_limits::max which implies it was not filled correctly.\nWhile this is "
+                 "valid for the matching variables as it can signal the failure to find a postive or negative seed "
+                 "trajectory,\nit is not valid for the layer number as that is always known.\nThus there is a logic "
+                 "error in the code or possible corupt data";
```
Comment (mmusich) [inline RecoEgamma/EgammaHLTProducers/plugins/EgammaHLTPixelMatchVarProducer.cc:53]:
```suggestion
                 "error in the code or possible corrupt data";
```

File: RecoEgamma/EgammaHLTProducers/plugins/EgammaHLTPixelMatchVarProducer.cc [hunk 2]
```diff
@@ -41,7 +41,24 @@ namespace {
       int layerOffset = 3;
       if (seed.subDet(hitNr) == PixelSubdetector::PixelEndcap)
         layerOffset += 4;
-      int layerBit = 0x1 << layerOffset << seed.layerOrDiskNr(hitNr);
+
+      int layerOrDiskNr = seed.layerOrDiskNr(hitNr);
+      if (layerOrDiskNr == 0 || layerOrDiskNr > 4) {
```
Comment (mmusich) [inline RecoEgamma/EgammaHLTProducers/plugins/EgammaHLTPixelMatchVarProducer.cc:46]: for my education what happens in the phase-2 case? You don't foresee to make use of the full FPix ?

Comment (Sam-Harper) [reply RecoEgamma/EgammaHLTProducers/plugins/EgammaHLTPixelMatchVarProducer.cc:46]: ah thanks, I had understood from our brief conversation that we were still 4 layers in phase-II and of course we dont, sorry my bad. 

Right now we dont take advantage of the full fpx only 1-4 and we should review this as it'll be important at high eta. But this requires a reworking of this code and more reason to have the warning in there to cause problems. 

Thanks for pointing it out

Comment (mmusich) [reply RecoEgamma/EgammaHLTProducers/plugins/EgammaHLTPixelMatchVarProducer.cc:46]: > ah thanks, I had understood from our brief conversation that we were still 4 layers in phase-II and of course we dont, sorry my bad.

In phase-2 we'll still have 4 layers in BPix, but 12 disks in the endcaps:

image

sorry it was probably me not being clear during our conversation.
So, IIUC, you don't think this 4 should be a 12 for the time being until there's more code rework, right?

**Comment by Sam-Harper** [comment]
@cmsbuild, please test

**Comment by Martin-Grunewald** [comment]
@Sam-Harper what's the status of this PR?

**Comment by Martin-Grunewald** [comment]
Thanks, Sam!

**Comment by Martin-Grunewald** [comment]
+1

**Comment by mandrenguyen** [comment]
+1

**Pull request merged**